sdk/keyvault/keyvault-certificates/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -0,0 +1,149 @@
+import { WebResource } from "@azure/core-http";
+
+export const TokenRefreshBufferMs = 2 * 60 * 1000; // 2 Minutes
+
+/**
+ * Creates a new ChallengeBasedAuthenticationPolicy factory.
+ *
+ * @param credential The TokenCredential implementation that can supply the challenge token.
+ */
+export function challengeBasedAuthenticationPolicy(credential: TokenCredential): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new ChallengeBasedAuthenticationPolicy(nextPolicy, options, credential);
+    }
+  };
+}
+
+export class AuthenticationChallenge {
+  constructor(public scopes: string[] | string) { }
+}
+
+/**
+ *
+ * Provides a RequestPolicy that can request a token from a TokenCredential
+ * implementation and then apply it to the Authorization header of a request
+ * as a Bearer token.
+ *
+ */
+export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+  private cachedToken: AccessToken | undefined = undefined;
+  private challenge: AuthenticationChallenge | undefined = undefined;
+
+  /**
+   * Creates a new ChallengeBasedAuthenticationPolicy object.
+   *
+   * @param nextPolicy The next RequestPolicy in the request pipeline.
+   * @param options Options for this RequestPolicy.
+   * @param credential The TokenCredential implementation that can supply the bearer token.
+   * @param scopes The scopes for which the bearer token applies.
+   */
+  constructor(
+    nextPolicy: RequestPolicy,
+    options: RequestPolicyOptions,
+    private credential: TokenCredential,
+    private refreshOn: number = Date.now(),
+  ) {
+    super(nextPolicy, options);
+  }
+
+  private parseWWWAuthenticate(www_authenticate: string): string {
+    // Parses an authentication message like:
+    // ```
+    // Bearer authorization="some_authorization", resource="https://some.url"
+    // ```
+    let authenticateArray = www_authenticate.split(" ");
+
+    // Remove the "Bearer" piece
+    delete authenticateArray[0];
+
+    // Split the KV comma-separated list
+    let commaSep = authenticateArray.join().split(",");
+    for (let item of commaSep) {
+      // Split the key/value pairs
+      let kv = item.split("=");
+      if (kv[0].trim() == "resource") {
+        // Remove the quotations around the string
+        let resource = kv[1].trim().replace(/['"]+/g, '');
+        return resource;
+      }
+    }
+    return "";
+  }
+
+  /**
+   * Applies the Bearer token to the request through the Authorization header.
+   * @param webResource
+   */
+  public async sendRequest(
+    webResource: WebResource
+  ): Promise<HttpOperationResponse> {
+    if (!webResource.headers) webResource.headers = new HttpHeaders();
+
+    let originalBody = webResource.body;
+
+    let headers = webResource.headers;
+    if (this.challenge == undefined) {
+      // Use a blank to start the challenge
+      webResource.headers = new HttpHeaders();
```

Comment:
>webResource.headers = new HttpHeaders(); [](start = 6, length = 40)

This should actually use the headers from the original request.  At the very least the Host header is needed to correctly identity the vault (different vaults may have different auth schemes), but it's safe just to use all the headers.